README.md
# Kik Python API

[![PyPI Version][pypi-image]][pypi-url] [![Docs Status][docs-image]][docs-url]

This library provides tools to build bots that communicate with the Kik API.
Find out more and create your bot at [dev.kik.com][dev-kik-url].

This package can be imported using pip by adding the following to your `requirements.txt`.

```
kik==<version>
```

# Docs

Our docs are available [on readthedocs][docs-url].

## Contributing

If you're looking to contribute the the `kik` package, check out the [Contributing Guide](/CONTRIBUTING.md).

This project adheres to the Contributor Covenant [code of conduct](CODE_OF_CONDUCT.md).
By participating, you are expected to uphold this code. Please report unacceptable behavior to bots@kik.com.

## License

The kik library is released under the terms of the MIT license.
See [License](LICENSE.md) for more information or see https://opensource.org/licenses/MIT.

[dev-kik-url]: https://dev.kik.com/

[docs-image]: https://readthedocs.org/projects/kik/badge/?version=latest
[docs-url]: https://kik.readthedocs.org

[pypi-image]: https://img.shields.io/pypi/v/kik.svg
[pypi-url]: https://pypi.python.org/pypi/kik


HISTORY.md
# Release History

## v.1.0.7
- Bugfix for KikApi.verify_signature in Python 3

## v.1.0.6
- Bugfix for setting configurations with no features
- Added configuration to the example bot

## v.1.0.5
- Fix installation error due to missing README

## v1.0.4
- Bugfix for Python 3 support
- Updated README

## v1.0.3
- Update setup.py so that PyPI will actually render it
- Fix error in configuration docs
- Standardize routes in documentation

## v1.0.2
- Switch to a Restructured Text Readme so PyPI can render it

## v1.0.0
- Initial Release


test/test_api.py
import json
from unittest import TestCase

import mock
import requests

from kik import KikApi, KikError, Code, Configuration, User
from kik.messages import TextMessage


def _response(status_code, content):
    response = requests.Response()
    response.status_code = status_code
    response._content = content
    return response


def requests_raise_func(method, url, *args, **kwargs):
    raise Exception('Must mock {} request to {}'.format(method.upper(), url))


class KikBotApiTest(TestCase):
    def setUp(self):
        super(KikBotApiTest, self).setUp()

        self.requests_mock = mock.patch('requests.api.request', wraps=requests_raise_func)
        self.requests_mock.start()

        self.api = KikApi('mybotusername', 'mybotapikey')

    def tearDown(self):
        super(KikBotApiTest, self).tearDown()

        self.requests_mock.stop()

    @mock.patch('requests.post', return_value=_response(200, json.dumps({}).encode('utf-8')))
    def test_send_messages(self, post):
        msgs = [TextMessage(to='aleem', body='Sometext')]

        response = self.api.send_messages(msgs)

        post_call = post.call_args
        self.assertEqual(post_call[0][0], 'https://api.kik.com/v1/message')
        self.assertEqual(post_call[1]['auth'], ('mybotusername', 'mybotapikey'))
        self.assertEqual(post_call[1]['timeout'], 60)
        self.assertEqual(post_call[1]['headers'], {'Content-Type': 'application/json'})
        self.assertEqual(json.loads(post_call[1]['data']), {
            'messages': [{
                'type': 'text',
                'to': 'aleem',
                'body': 'Sometext',
            }]
        })
        self.assertEqual(response, {})

    @mock.patch('requests.post', return_value=_response(200, json.dumps({}).encode('utf-8')))
    def test_broadcast_messages(self, post):
        msgs = [TextMessage(to='aleem', body='Sometext')]

        response = self.api.send_broadcast(msgs)

        post_call = post.call_args
        self.assertEqual(post_call[0][0], 'https://api.kik.com/v1/broadcast')
        self.assertEqual(post_call[1]['auth'], ('mybotusername', 'mybotapikey'))
        self.assertEqual(post_call[1]['timeout'], 60)
        self.assertEqual(post_call[1]['headers'], {'Content-Type': 'application/json'})
        self.assertEqual(json.loads(post_call[1]['data']), {
            'messages': [{
                'type': 'text',
                'to': 'aleem',
                'body': 'Sometext'
            }]
        })
        self.assertEqual(response, {})

    @mock.patch('requests.post', return_value=_response(400, json.dumps({'error': 'BadRequest'}).encode('utf-8')))
    def test_send_messages_failure(self, post):
        msgs = [TextMessage(to='aleem', body='Sometext')]

        self.assertRaises(KikError, self.api.send_messages, msgs)

    @mock.patch('requests.post', return_value=_response(400, json.dumps({'error': 'BadRequest'}).encode('utf-8')))
    def test_send_broadcast_failure(self, post):
        msgs = [TextMessage(to='aleem', body='Sometext')]

        self.assertRaises(KikError, self.api.send_broadcast, msgs)

    @mock.patch('requests.get', return_value=_response(200, json.dumps({
        'firstName': 'First',
        'lastName': 'Last',
        'profilePicUrl': 'http://foo.bar/profile',
        'profilePicLastModified': 1458657367
    }).encode('utf-8')))
    def test_get_user_profile(self, get):
        user = self.api.get_user('aleem')

        get.assert_called_once_with(
            'https://api.kik.com/v1/user/aleem',
            auth=('mybotusername', 'mybotapikey'),
            timeout=60
        )

        self.assertIsInstance(user, User)
        self.assertEqual(user.first_name, 'First')
        self.assertEqual(user.last_name, 'Last')
        self.assertEqual(user.profile_pic_url, 'http://foo.bar/profile')
        self.assertEqual(user.profile_pic_last_modified, 1458657367)

    @mock.patch('requests.get', return_value=_response(400, json.dumps({'error': 'BadRequest'}).encode('utf-8')))
    def test_get_user_profile_failure(self, get):
        self.assertRaises(KikError, self.api.get_user, 'aleem')

    @mock.patch('requests.post', return_value=_response(
        200, json.dumps({'id': 'ba7a319394f912ccad1ac42770529bd5cb0e9783'}).encode('utf-8')
    ))
    def test_create_kik_code(self, post):
        code = self.api.create_code({'akey': 'avalue'})

        post.assert_called_once_with(
            'https://api.kik.com/v1/code',
            timeout=60,
            auth=('mybotusername', 'mybotapikey'),
            headers={
                'Content-Type': 'application/json'
            },
            data=json.dumps({'data': '{"akey": "avalue"}'})
        )

        self.assertIsInstance(code, Code)
        self.assertEqual(code.id, 'ba7a319394f912ccad1ac42770529bd5cb0e9783')

    @mock.patch('requests.post', return_value=_response(
        200, json.dumps({'id': 'ba7a319394f912ccad1ac42770529bd5cb0e9783'}).encode('utf-8')
    ))
    def test_create_kik_code_no_data(self, post):
        code = self.api.create_code()

        post.assert_called_once_with(
            'https://api.kik.com/v1/code',
            timeout=60,
            auth=('mybotusername', 'mybotapikey'),
            headers={
                'Content-Type': 'application/json'
            },
            data='{}'
        )

        self.assertIsInstance(code, Code)
        self.assertEqual(code.id, 'ba7a319394f912ccad1ac42770529bd5cb0e9783')

    @mock.patch('requests.post', return_value=_response(400, json.dumps({'error': 'BadRequest'}).encode('utf-8')))
    def test_create_kik_code_failure(self, post):
        self.assertRaises(KikError, self.api.create_code, {'akey': 'avalue'})

    @mock.patch('requests.get', return_value=_response(200, json.dumps({
        'webhook': 'https://example.com/incoming',
        'features': {
            'manuallySendReadReceipts': True
        }
    }).encode('utf-8')))
    def test_get_configuration(self, get):
        config = self.api.get_configuration()

        get.assert_called_once_with(
            'https://api.kik.com/v1/config',
            timeout=60,
            auth=('mybotusername', 'mybotapikey')
        )

        self.assertEqual(config.webhook, 'https://example.com/incoming')
        self.assertEqual(config.features, {'manuallySendReadReceipts': True})

    @mock.patch('requests.post', return_value=_response(200, json.dumps({
        'webhook': 'https://example.com/incoming',
        'features': {
            'manuallySendReadReceipts': True
        }
    }).encode('utf-8')))
    def test_set_configuration(self, post):
        config = Configuration(
            webhook='https://example.com/incoming',
            features={'manuallySendReadReceipts': True}
        )

        response = self.api.set_configuration(config)

        self.assertEqual(post.call_count, 1)
        self.assertEqual(post.call_args[0][0], 'https://api.kik.com/v1/config')
        self.assertEqual(post.call_args[1]['timeout'], 60)
        self.assertEqual(post.call_args[1]['auth'], ('mybotusername', 'mybotapikey'))
        self.assertEqual(post.call_args[1]['headers'], {'Content-Type': 'application/json'})
        self.assertEqual(json.loads(post.call_args[1]['data']), {
            'webhook': 'https://example.com/incoming',
            'features': {
                'manuallySendReadReceipts': True
            }
        })

        self.assertIsInstance(response, Configuration)
        self.assertEqual(response.webhook, 'https://example.com/incoming')
        self.assertEqual(response.features, {'manuallySendReadReceipts': True})

    def test_verify_signature(self):
        self.assertTrue(self.api.verify_signature('AC18D0105C2C257652859322B0499313342C6EB9', b'body'))
        self.assertFalse(self.api.verify_signature('fakesig', b'body'))


test/test_code.py
from unittest import TestCase

from kik import Code


class KikCodeTest(TestCase):
    def test_from_json(self):
        code = Code.from_json({'id': 'ba7a319394f912ccad1ac42770529bd5cb0e9783'})
        self.assertEqual(code.id, 'ba7a319394f912ccad1ac42770529bd5cb0e9783')

    def test_url(self):
        code = Code('ba7a319394f912ccad1ac42770529bd5cb0e9783')
        self.assertEqual(code.url(), 'https://api.kik.com/v1/code/ba7a319394f912ccad1ac42770529bd5cb0e9783')
        self.assertEqual(code.url(color=Code.Colors.TURQUOISE),
                         'https://api.kik.com/v1/code/ba7a319394f912ccad1ac42770529bd5cb0e9783?c=1')


test/test_user.py
from unittest import TestCase

from kik.user import User


class UserTest(TestCase):
    def test_from_json(self):
        user = User.from_json({
            'firstName': 'First',
            'lastName': 'Last',
            'profilePicUrl': 'http://foo.bar/profile',
            'profilePicLastModified': 1458657367
        })
        self.assertEqual(user.first_name, 'First')
        self.assertEqual(user.last_name, 'Last')
        self.assertEqual(user.profile_pic_url, 'http://foo.bar/profile')
        self.assertEqual(user.profile_pic_last_modified, 1458657367)

    def test_from_json_no_picture(self):
        user = User.from_json({
            'firstName': 'First',
            'lastName': 'Last',
            'profilePicUrl': None,
            'profilePicLastModified': None
        })
        self.assertEqual(user.first_name, 'First')
        self.assertEqual(user.last_name, 'Last')
        self.assertEqual(user.profile_pic_url, None)
        self.assertEqual(user.profile_pic_last_modified, None)


kik/api.py
import base64
import hashlib
import hmac
import json

import requests

from kik.code import Code
from kik.configuration import Configuration
from kik.error import KikError
from kik.user import User

ROOT_URL = 'https://api.kik.com{}'


class KikApi(object):
    """
    Generic API Client for all Kik API features.

    :param bot: Your bot's username
    :type bot: str
    :param api_key: Your bot's API key
    :type api_key: str
    """
    def __init__(self, bot, api_key):
        super(KikApi, self).__init__()
        self.bot = bot
        self.api_key = api_key

    def send_messages(self, messages):
        """
        Sends a batch of messages.

        :param messages: List of :class:`Message <kik.messages.Message>` to be sent.
        :type messages: list[kik.messages.Message]
        :returns: A dict containing the response from the API
        :rtype: dict

        .. note:: Subject to limits on the number of messages sent, documented at
            `<https://dev.kik.com/#/docs/messaging#sending-messages>`_.

        Usage:

        >>> from kik import KikApi
        >>> from kik.messages import TextMessage
        >>> kik = KikApi(BOT_USERNAME, BOT_API_KEY)
        >>> kik.send_messages([
        >>>     TextMessage(
        >>>         to='ausername',
        >>>         chat_id='2e566cf66b07d9622053b2f0e44dd14926d89a6d61adf496844781876d62cca6',
        >>>         body='Some Text'
        >>>     )
        >>> ])
        {}
        """
        response = requests.post(
            ROOT_URL.format('/v1/message'),
            auth=(self.bot, self.api_key),
            timeout=60,
            headers={
                'Content-Type': 'application/json'
            },
            data=json.dumps({'messages': [m.to_json() for m in messages]})
        )

        if response.status_code != 200:
            raise KikError(response.text)

        return response.json()

    def send_broadcast(self, messages):
        """
        Sends a batch of messages though the broadcast API.

        :param messages: List of :class:`Message <kik.messages.message.Message>` to be sent.
        :type messages: list[kik.messages.Message]
        :returns: A dict containing the response from the API.
        :rtype: dict

        .. note:: Subject to limits on the number of messages sent, documented at
            `<https://dev.kik.com/#/docs/messaging#broadcasting>`_.

        Usage:

        >>> from kik import KikApi
        >>> from kik.messages import TextMessage
        >>> kik = KikApi(BOT_USERNAME, BOT_API_KEY)
        >>> kik.send_broadcast([
        >>>     TextMessage(
        >>>         to='ausername',
        >>>         chat_id='2e566cf66b07d9622053b2f0e44dd14926d89a6d61adf496844781876d62cca6',
        >>>         body='Some Text'
        >>>     )
        >>> ])
        {}
        """
        response = requests.post(
            ROOT_URL.format('/v1/broadcast'),
            auth=(self.bot, self.api_key),
            timeout=60,
            headers={
                'Content-Type': 'application/json'
            },
            data=json.dumps({'messages': [m.to_json() for m in messages]})
        )

        if response.status_code != 200:
            raise KikError(response.text)

        return response.json()

    def get_user(self, username):
        """
        Gets a user's profile data.

        :param username: List of :class:`Message <kik.messages.Message>` to be sent.
        :type username: str
        :returns: A :class:`User <kik.User>` containing the user's profile data.
        :rtype: kik.User

        .. note:: In order to fetch a user's profile, the user must be a subscriber to your bot

        Usage:

        >>> from kik import KikApi
        >>> kik = KikApi(BOT_USERNAME, BOT_API_KEY)
        >>> kik.get_user('aleem')
        <kik.User>
        """
        response = requests.get(
            ROOT_URL.format('/v1/user/{}'.format(username)),
            auth=(self.bot, self.api_key),
            timeout=60
        )

        if response.status_code != 200:
            raise KikError(response.text)

        content = response.json()

        return User.from_json(content)

    def create_code(self, data=None):
        """
        Creates a Kik Code for your bot.

        :param data: (optional) Data to embed in the code, which will be returned in a `scan-data` message when the
            code is scanned.
        :type data: dict, str
        :returns: A :class:`Code <kik.Code>` representing the code.
        :rtype: kik.Code

        Usage:

        >>> from kik import KikApi
        >>> kik = KikApi(BOT_USERNAME, BOT_API_KEY)
        >>> kik.create_code(data='somedata')
        <kik.Code>
        """
        if data is not None:
            payload = json.dumps({'data': json.dumps(data) if isinstance(data, dict) else data})
        else:
            payload = json.dumps({})

        response = requests.post(
            ROOT_URL.format('/v1/code'),
            auth=(self.bot, self.api_key),
            headers={
                'Content-Type': 'application/json'
            },
            data=payload,
            timeout=60
        )

        if response.status_code != 200:
            raise KikError(response.text)

        content = response.json()

        return Code.from_json(content)

    def get_configuration(self):
        """
        Retrieves your bot's configuration

        :returns: A :class:`Configuration <kik.Configuration>` representing the configuration data.
        :rtype: kik.Configuration

        Usage:

        >>> from kik import KikApi
        >>> kik = KikApi(BOT_USERNAME, BOT_API_KEY)
        >>> kik.get_configuration()
        <kik.Configuration>
        """
        response = requests.get(
            ROOT_URL.format('/v1/config'),
            auth=(self.bot, self.api_key),
            timeout=60
        )

        if response.status_code != 200:
            raise KikError(response.text)

        content = response.json()

        return Configuration.from_json(content)

    def set_configuration(self, config):
        """
        Sets your bot's configuration

        :param config: A :class:`Configuration<kik.Configuration>` containing your bot's new configuration
        :type config: kik.Configuration
        :returns: A dict containing the response from the API.
        :rtype: dict

        Usage:

        >>> from kik import KikApi, Configuration
        >>> kik = KikApi(BOT_USERNAME, BOT_API_KEY)
        >>> config = Configuration(webhook='https://example.com/incoming')
        >>> kik.set_configuration(config)
        {}
        """
        response = requests.post(
            ROOT_URL.format('/v1/config'),
            auth=(self.bot, self.api_key),
            headers={
                'Content-Type': 'application/json'
            },
            data=json.dumps(config.to_json()),
            timeout=60
        )

        if response.status_code != 200:
            raise KikError(response.text)

        return response.json()

    def verify_signature(self, signature, body):
        """
        Verifies that a request body correctly matches the header signature.
        For more on signatures see `<https://dev.kik.com/#/docs/messaging#receiving-messages>`_.

        In Python 3, `body` must be a bytestring
        """

        expected = base64.b16encode(hmac.new(self.api_key.encode('utf-8'), body, hashlib.sha1).digest()).decode('utf-8')

        return signature == expected
